File: python3/leetcodepy/scramble_string.py
"""
We can scramble a string s to get a string t using the following algorithm:

1. If the length of the string is 1, stop.
2. If the length of the string is > 1, do the following:
    Split the string into two non-empty substrings at a random index, i.e., if the string is s, divide it to x and y where s = x + y.
    Randomly decide to swap the two substrings or to keep them in the same order. i.e., after this step, s may become s = x + y or s = y + x.
    Apply step 1 recursively on each of the two substrings x and y.

Given two strings s1 and s2 of the same length, return true if s2 is a scrambled string of s1, otherwise, return false.



Example 1:

Input: s1 = "great", s2 = "rgeat"
Output: true
Explanation: One possible scenario applied on s1 is:
"great" --> "gr/eat" // divide at random index.
"gr/eat" --> "gr/eat" // random decision is not to swap the two substrings and keep them in order.
"gr/eat" --> "g/r / e/at" // apply the same algorithm recursively on both substrings. divide at ranom index each of them.
"g/r / e/at" --> "r/g / e/at" // random decision was to swap the first substring and to keep the second substring in the same order.
"r/g / e/at" --> "r/g / e/ a/t" // again apply the algorithm recursively, divide "at" to "a/t".
"r/g / e/ a/t" --> "r/g / e/ a/t" // random decision is to keep both substrings in the same order.
The algorithm stops now and the result string is "rgeat" which is s2.
As there is one possible scenario that led s1 to be scrambled to s2, we return true.

Example 2:

Input: s1 = "abcde", s2 = "caebd"
Output: false

Example 3:

Input: s1 = "a", s2 = "a"
Output: true


Constraints:

s1.length == s2.length
1 <= s1.length <= 30
s1 and s2 consist of lower-case English letters.
"""


class Solution1:
    def isScramble(self, s1: str, s2: str) -> bool:
        return self._is_scramble(s1, s2, {})

    def _is_scramble(self, s1: str, s2: str, _dict: dict) -> bool:
        k = s1 + s2
        if k in _dict:
            return _dict[k]
        _len = len(s1)
        if _len == 1:
            _dict[k] = s1 == s2
            return s1 == s2
        if sorted(s1) != sorted(s2):
            _dict[k] = False
            return False
        for l in range(1, _len):
            if (self._is_scramble(s1[:l], s2[:l], _dict) and self._is_scramble(s1[l:_len], s2[l:_len], _dict)) or \
                    (self._is_scramble(s1[:l], s2[_len - l:_len], _dict) and self._is_scramble(s1[l:_len],
                                                                                               s2[:_len - l], _dict)):
                _dict[k] = True
                return True
        _dict[k] = False
        return False


class Solution2:
    def isScramble(self, s1: str, s2: str) -> bool:
        size = len(s1)
        dp = [[[False] * size for _ in range(size)] for __ in range(size + 1)]
        for i in range(size):
            for j in range(size):
                if s1[i] == s2[j]:
                    dp[1][i][j] = True
        for length in range(2, size + 1):
            for i in range(size - length + 1):
                for j in range(size - length + 1):
                    for len_ in range(1, length):
                        if dp[len_][i][j] and dp[length - len_][i + len_][j + len_] \
                                or dp[len_][i][j + length - len_] and dp[length - len_][i + len_][j]:
                            dp[length][i][j] = True
                            break
        return dp[size][0][0]

File: java/src/main/java/io/binac/leetcode/ScrambleString.java
package io.binac.leetcode;

import javafx.util.Pair;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * <p>We can scramble a string s to get a string t using the following algorithm:</p>
 *
 * <ol>
 * 	<li>If the length of the string is 1, stop.</li>
 * 	<li>If the length of the string is &gt; 1, do the following:
 * 	<ul>
 * 		<li>Split the string into two non-empty substrings at a random index, i.e., if the string is <code>s</code>, divide it to <code>x</code> and <code>y</code> where <code>s = x + y</code>.</li>
 * 		<li><strong>Randomly</strong>&nbsp;decide to swap the two substrings or to keep them in the same order. i.e., after this step, <code>s</code> may become <code>s = x + y</code> or <code>s = y + x</code>.</li>
 * 		<li>Apply step 1 recursively on each of the two substrings <code>x</code> and <code>y</code>.</li>
 * 	</ul>
 * 	</li>
 * </ol>
 *
 * <p>Given two strings <code>s1</code> and <code>s2</code> of <strong>the same length</strong>, return <code>true</code> if <code>s2</code> is a scrambled string of <code>s1</code>, otherwise, return <code>false</code>.</p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 *
 * <pre><strong>Input:</strong> s1 = "great", s2 = "rgeat"
 * <strong>Output:</strong> true
 * <strong>Explanation:</strong> One possible scenario applied on s1 is:
 * "great" --&gt; "gr/eat" // divide at random index.
 * "gr/eat" --&gt; "gr/eat" // random decision is not to swap the two substrings and keep them in order.
 * "gr/eat" --&gt; "g/r / e/at" // apply the same algorithm recursively on both substrings. divide at ranom index each of them.
 * "g/r / e/at" --&gt; "r/g / e/at" // random decision was to swap the first substring and to keep the second substring in the same order.
 * "r/g / e/at" --&gt; "r/g / e/ a/t" // again apply the algorithm recursively, divide "at" to "a/t".
 * "r/g / e/ a/t" --&gt; "r/g / e/ a/t" // random decision is to keep both substrings in the same order.
 * The algorithm stops now and the result string is "rgeat" which is s2.
 * As there is one possible scenario that led s1 to be scrambled to s2, we return true.
 * </pre>
 *
 * <p><strong>Example 2:</strong></p>
 *
 * <pre><strong>Input:</strong> s1 = "abcde", s2 = "caebd"
 * <strong>Output:</strong> false
 * </pre>
 *
 * <p><strong>Example 3:</strong></p>
 *
 * <pre><strong>Input:</strong> s1 = "a", s2 = "a"
 * <strong>Output:</strong> true
 * </pre>
 *
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li><code>s1.length == s2.length</code></li>
 * 	<li><code>1 &lt;= s1.length &lt;= 30</code></li>
 * 	<li><code>s1</code> and <code>s2</code> consist of lower-case English letters.</li>
 * </ul>
 */
public class ScrambleString {
    public static class Solution1 {
        private final int[] EQ_COUNT = new int[26];

        private boolean isScramble(String s1, String s2, Map<String, Boolean> map) {
            String k = s1 + s2;
            if (map.containsKey(k))
                return map.get(k);
            final int len = s1.length();
            if (len == 1) {
                boolean b = s1.equals(s2);
                map.put(k, b);
                return b;
            }
            int[] count = new int[26];
            for (int i = 0; i < len; ++i) {
                ++count[s1.charAt(i) - 'a'];
                --count[s2.charAt(i) - 'a'];
            }
            if (!Arrays.equals(count, EQ_COUNT)) {
                map.put(k, false);
                return false;
            }
            for (int l = 1; l < len; ++l) {
                if ((isScramble(s1.substring(0, l), s2.substring(0, l), map) && isScramble(s1.substring(l, len), s2.substring(l, len), map))
                        || (isScramble(s1.substring(0, l), s2.substring(len - l, len), map) && isScramble(s1.substring(l, len), s2.substring(0, len - l), map))) {
                    map.put(k, true);
                    return true;
                }
            }
            map.put(k, false);
            return false;
        }

        public boolean isScramble(String s1, String s2) {
            return isScramble(s1, s2, new HashMap<>());
        }
    }

    public static class Solution2 {
        public boolean isScramble(String s1, String s2) {
            final int len = s1.length();

            boolean[][][] dp = new boolean[len + 1][len][len];
            for (int i = 0, j; i < len; ++i) {
                for (j = 0; j < len; ++j) {
                    if (s1.charAt(i) == s2.charAt(j))
                        dp[1][i][j] = true;
                }
            }
            for (int l = 2, i, j, k; l <= len; ++l) {
                for (i = 0; i + l <= len; ++i) {
                    for (j = 0; j + l <= len; ++j) {
                        for (k = 1; k < l; ++k) {
                            if ((dp[k][i][j] && dp[l - k][i + k][j + k])
                                    || (dp[k][i][j + l - k] && dp[l - k][i + k][j])) {
                                dp[l][i][j] = true;
                                break;
                            }
                        }
                    }
                }
            }
            return dp[len][0][0];
        }
    }
}
